Automatically generated goal names (#20809)
rocq-prover/rocq: doc/sphinx/proofs/writing-proofs/proof-mode.rst

   - "Strict Subproofs": this makes bullets active (this is the default behavior).

.. _named_goals:

Named goals
~~~~~~~~~~~

You can focus on a goal by using its name. Goals do not have a name by default,
but a name can be given by using :n:`refine ?[@ident]`, or generated using the
:flag:`Generate Goal Names` flag.

.. flag:: Generate Goal Names

   Enables automatic generation of goal names for the :tacn:`induction`,
   :tacn:`destruct` and :tacn:`eapply` tactics. For :tacn:`induction` and
   :tacn:`destruct`, the subgoal takes the name of the corresponding
   constructor. For :tacn:`eapply`, the subgoal takes the name of the
   corresponding hypothesis.

   This option makes it possible to write proofs with multiple subgoals that do
   not depend on the order in which constructors were defined, but instead rely
   on the constructor names. If you use bullets or numbers, reordering
   constructors will break the proof.

   For proofs that use nested :tacn:`induction` or case analysis, qualified
   names such as `true.false` are used to disambiguate subgoals (see an example
   :ref:`here <qualified-goal-names>`).

   .. example:: Automatic generation of goal names

>       For the example given :ref:`here <example-working-with-named-goals>`,
>       names are generated for both the base case and the induction case.

Reviewer: This wording suggests that the reader should immediately look at the other example (ie distracting).

```suggestion
      This example uses automatically generated goal names for both the base
      and induction cases, in contrast to the example
      :ref:`here <example-working-with-named-goals>`,
      in which the user explicitly assigns goal names with :cmd:`refine`.
```

      .. rocqtop:: in

         Set Generate Goal Names.
         Goal forall n, n + 0 = n.

      .. rocqtop:: all

         induction n.
         [O]: { (* O and S are the constructors for nat. *)
           reflexivity.

      .. rocqtop:: in abort

         }

      If a goal comes from a variable that failed to instantiate (e.g. when
      using :tacn:`eapply` or other `e*` tactics), the goal is named after the

>       variable:

Reviewer: should use the term "existential variable" here.  "failed to instantiate" is needlessly vague.  Mention that `?Rwf` and `?R`come from the definition of `well_founded_ind`.

      .. rocqtop:: none

         Inductive even : nat -> Prop :=
         | even_zero : even 0
         | even_succ : forall n, even n -> even (S (S n)).

         Inductive odd : nat -> Prop :=
         | odd_one : odd 1
         | odd_succ : forall n, odd n -> odd (S (S n)).

      .. rocqtop:: in

         Goal forall n : nat, even n \/ odd n.

      .. rocqtop:: all abort

         eapply well_founded_ind.

>          [R]: exact lt.

Reviewer: For usability, the goal display should show `goal 2 (?R) is:` on the second goal.  Maybe you can tweak the goal printing to do that (should be a simple change).

<img width="1142" height="461" alt="image" src="https://github.com/user-attachments/assets/8a5a3f2f-5f79-46ec-b977-618199268abf" />

Author: The second goal is not `?R`, but the inductive case. `?R` is shelved so it does not appear here.

   .. _qualified-goal-names:
   .. example:: Qualified goal names

      When doing nested case analysis or induction, qualified names are used to

>       disambiguate subgoals. The name of each subgoal is prefixed by the name of
>       its parent.

Reviewer: ```suggestion
      disambiguate subgoals. Subgoal names include the names of parent goals as a prefix.
```

   - "None": this makes bullets inactive.
   - "Strict Subproofs": this makes bullets active (this is the default behavior).

.. _named_goals:

Named goals
~~~~~~~~~~~

You can focus on a goal by using its name. Goals do not have a name by default,
but a name can be given by using :n:`refine ?[@ident]`, or generated using the
:flag:`Accessible Goal Names` flag.

.. flag:: Accessible Goal Names

   Enable automatic generation of goal names for the :tacn:`induction`,

>    :tacn:`destruct` and :tacn:`eapply` tactics. The goal generated takes the

Reviewer: ```suggestion
   :tacn:`destruct` and :tacn:`eapply` tactics. The goal name is
```

>    name of the constructor or a matching hypothesis for the goal name.

Author: This is very poorly worded, apologies. What I meant is that the goal name is the constructor name when doing `destruct`/`induction`, or a hypothesis name when doing `eapply`.

   .. example:: Automatic generation of goal names

      Continuing the example from :ref:`here <example-working-with-named-goals>`,
      names are generated for both the base case and the induction case.

      .. rocqtop:: all

         Set Accessible Goal Names.

         Goal forall n, n + 0 = n.
         Proof.
         induction n.
         [O]: { (* O is the name of the constructor for zero. *)
           reflexivity.

      .. rocqtop:: in

         }

      This also gives a name to goals that come from a binder or hypothesis:

      .. rocqtop:: all reset

         Set Accessible Goal Names.

         Goal exists n : nat, n = n.
         eexists.
         [n]: exact 0.
         reflexivity.
         Qed.

   .. example:: Conflicting names

      If several goals generate the same name (e.g. when doing nested case
      analysis or induction), *qualified names* are used to disambiguate them,
      using the qualified name of the parent as the basis for the fully
      qualified name.

      .. rocqtop:: all abort

         Set Accessible Goal Names.

         Goal forall n m, n + m = m + n.
         Proof.
         intros. induction m.
         [O]: {

>            simpl. induction n.

Author: Should be working now.

   - "None": this makes bullets inactive.
   - "Strict Subproofs": this makes bullets active (this is the default behavior).

.. _named_goals:

Named goals
~~~~~~~~~~~

You can focus on a goal by using its name. Goals do not have a name by default,
but a name can be given by using :n:`refine ?[@ident]`, or generated using the
:flag:`Generate Goal Names` flag.

.. flag:: Generate Goal Names

   Enable automatic generation of goal names for the :tacn:`induction`,
   :tacn:`destruct` and :tacn:`eapply` tactics. For :tacn:`induction` and
   :tacn:`destruct`, the subgoal takes the name of the corresponding
   constructor. For :tacn:`eapply`, the subgoal takes the name of the
   corresponding hypothesis.

   This option makes it possible to write proofs with multiple subgoals that do
   not depend on the order in which constructors were defined. If you use

>    bullets or numbers instead, reordering constructors will break the proof.

Reviewer: Also add a sentence mentioning qualified names here with an in-line example like `O.S` and a hyperlink to the longer example.  It's a key point that should be in the text description, not only in the examples a couple pages below.

Author: I've added an example that uses nested focusing braces, so that should be clearer.

>    bullets or numbers instead, reordering constructors will break the proof.

Author: Yes indeed, might be worth adding to the documentation

   .. example:: Automatic generation of goal names

      Continuing the example from :ref:`here <example-working-with-named-goals>`,
      names are generated for both the base case and the induction case.

      .. rocqtop:: all

         Set Generate Goal Names.

         Goal forall n, n + 0 = n.
         Proof.
         induction n.
         [O]: { (* O is the name of the constructor for zero. *)
           reflexivity.

      .. rocqtop:: in

         }

      This also gives a name to goals that come from a binder or hypothesis:

      .. rocqtop:: all reset

>          Set Generate Goal Names.

Author: I don't know yet what will be the default value of the option yet, so I keep it in the examples for now

         Goal exists n : nat, n = n.
         eexists.
         [n]: exact 0.
         reflexivity.
         Qed.

   .. example:: Conflicting names

      If several goals generate the same name (e.g. when doing nested case
      analysis or induction), *qualified names* are used to disambiguate them,
      using the qualified name of the parent as the basis for the fully

>       qualified name.

Reviewer: If you have induction on a bool inside induction on a nat, will the bool cases have names like `true` or `O.true`?
I suspect that qualified names are always used for nested cases, but I didn't try an example.

Seems like you should have an example that actually selects a qualified goal name, e.g. `[S.O]: {`

      .. rocqtop:: all abort

         Set Generate Goal Names.

         Goal forall n m, n + m = m + n.
         Proof.
         intros. induction m.
         [O]: {

>            simpl. induction n.

Author: > There seems to be a bug in the rocqtop directive: the output for [O]: { should appear just below that line.

This is also present in other examples, so it might be a parsing issue.

I'll shorten the output a bit :+1:

>            simpl. induction n.

Reviewer: It's an bug/limitation in the Python code for the directive.

>            simpl. induction n.

Reviewer: >> it would be nicer if the user could specify only the last component of the goal name

> I do agree but goal names are global to the proof and thus only specifying `S` would be ambiguous, which is precisely why qualified names are required.

Just check at run time whether there are multiple goal names that match `S` and, if so, generate an error.  If your trie stores the qualified name in reverse order, that should be easy to do.  (Similar to qualids for modules: the user generally only needs to provide a suffix that uniquely identifies the module.)

   - "None": this makes bullets inactive.
   - "Strict Subproofs": this makes bullets active (this is the default behavior).

.. _named_goals:

Named goals
~~~~~~~~~~~

You can focus on a goal by using its name. Goals do not have a name by default,

> but a name can be given by using :n:`refine ?[@ident]`, or generated using the

Reviewer: ```suggestion
but a name can be given by using :n:`refine ?[@qualid]`, or generated using the
```

If correct.

Author: That is not allowed since qualified names are only used for disambiguation purposes.

:flag:`Generate Goal Names` flag.

.. flag:: Generate Goal Names

>    Enable automatic generation of goal names for the :tacn:`induction`,

Reviewer: ```suggestion
   Enables automatic generation of goal names for the :tacn:`induction`,
```

   :tacn:`destruct` and :tacn:`eapply` tactics. For :tacn:`induction` and
   :tacn:`destruct`, the subgoal takes the name of the corresponding
   constructor. For :tacn:`eapply`, the subgoal takes the name of the
   corresponding hypothesis.

   This option makes it possible to write proofs with multiple subgoals that do
   not depend on the order in which constructors were defined, but instead rely
   on the constructor names. If you use bullets or numbers, reordering
   constructors will break the proof.

   For proofs that use nested :tacn:`induction` or case analysis, qualified
   names such as `true.false` are used to disambiguate subgoals (see an example
   :ref:`here <qualified-goal-names>`).

   .. example:: Automatic generation of goal names

      For the example given :ref:`here <example-working-with-named-goals>`,
      names are generated for both the base case and the induction case.

>       .. rocqtop:: all
> 
>          Set Generate Goal Names.
> 
>          Goal forall n, n + 0 = n.

Reviewer: ```suggestion
      .. rocqtop:: in

         Set Generate Goal Names.

         Goal forall n, n + 0 = n.
         
      .. rocqtop:: all
```

No need to show the output for Goal--not interesting, more stuff for the reader to wade through.

         induction n.

>          [O]: { (* O is the name of the constructor for zero. *)

Reviewer: ```suggestion
         [O]: { (* O and S are the constructors for nat. *)
```

      .. rocqtop:: in abort

           reflexivity.
         }

>       This flag also gives a name to goals that come from a binder or hypothesis:

Reviewer: "binder" is not the right term here.  You should refer to the creation of an existential variable.

You don't give an example for hypotheses.  Offhand, I can't imagine how that would work.

Author: I've changed the example to show how goal names are generated for `eapply`, and removed the one with `eexists` since I believe the new example is better.

      .. rocqtop:: all reset

         Set Generate Goal Names.

         Goal exists n : nat, n = n.
         eexists.
         [n]: exact 0.
         reflexivity.
         Qed.

   .. _qualified-goal-names:
   .. example:: Qualified goal names

      When doing nested case analysis or induction, qualified names are used to
      disambiguate subgoals. The name of each subgoal is prefixed by the name of
      its parent.

>       .. rocqtop:: all

Reviewer: ```suggestion
      .. rocqtop:: in
```

This example becomes almost two screenfuls (tedious for the user to scan to figure out what parts are relevant)--much more understandable if you don't show all the output.  Leave it as an exercise for the reader to try the example.
